### ISSUE TITLE
Unittest for float serialization (testStreamRead) fails on i386 when compiling Minetest with GCC

### ISSUE THREAD

**Issue by erlehmann** [minetest/minetest#11810]
##### Minetest version
commit 80c3c7e642749f9316a3eee2c235df3ce8be1666

```
Minetest 5.5.0-dev-80c3c7e64 (Linux)
Using Lua 5.1.5
BUILD_TYPE=Release
RUN_IN_PLACE=0
USE_CURL=0
STATIC_SHAREDIR="/usr/local/share/minetest"
```

##### OS / Hardware
Operating system: Debian GNU/Linux i386

##### Summary
```
Test assertion failed: readF1000(is) == 53.534f
    at test_serialization.cpp:233
[FAIL] testStreamRead - 0ms
```

**Comment by sfan5** [comment]
Works on CI and cannot reproduce locally.

**Comment by neoh4x0r** [comment]
@erlehmann Are you sure that 1dc1305ad is the correct commit? 
That commit is for v5.5.0 -- but your minetest version says 5.4.1

PS: I just built it and  all tests passed after running **minetest --run-unitests**

**Comment by erlehmann** [comment]
@neoh4x0r yes I am pretty sure that Minetest 5.4.1 could not be built with Irrlicht 1.9.0mt3 and that it is a version number bug. I also could verify that it is new easily, because recent Minetest breaks a self test on Mineclonia startup due to this PR: https://github.com/minetest/minetest/pull/11770

I have rebuilt the entire thing to have an UBsan build it still claims to be 5.4.1 and still fails the unit test.

**Comment by neoh4x0r** [comment]
@erlehmann Does it still fail (and/or still report 5.4.1) if you build minetest according to the official readme (eg. not a UBsan build -- not sure how that is done anyway)?
https://github.com/minetest/minetest#build

**Comment by erlehmann** [comment]
@neoh4x0r as far as I know there is no “official” build skript that produces a binary reproducibly. I made my first build with:

```
cmake -B 5.4.0-447-g1dc1305ad && cd 5.4.0-447-g1dc1305ad && make -j2
```

I made my second build with:

```
cmake '-DCMAKE_CXX_FLAGS=-fsanitize=undefined -fno-sanitize=vptr' -B 5.4.0-447-g1dc1305ad && cd 5.4.0-447-g1dc1305ad && make -j2
```

**Comment by neoh4x0r** [comment]
@erlehmann The official way to build is the following:

```
$ git clone https://github.com/minetest/minetest
$ cd minetest

# checkout a specific commit, branch, or tag
$ git checkout 1dc1305ada07da8c2a278b46a34d58af86184af9

$ cmake -S . -B minetest/build/linux
$ make
$ ./bin/minetest --run-unitests
```

**Comment by erlehmann** [comment]
@neoh4x0r thank you. I plan to do it some time later, not today, because builds take a long time and make my computer a bit hot. Then I will report back.

**Comment by erlehmann** [comment]
During a build, I saw these warnings, which show how the version number discrepancy occurs:
```
[ 38%] Building CXX object src/CMakeFiles/minetest.dir/util/areastore.cpp.o
In file included from src/config.h:13,
                 from src/util/serialize.h:27,
                 from src/util/areastore.cpp:21:
src/cmake_config.h:8: warning: "VERSION_MINOR" redefined
 #define VERSION_MINOR 4
 
In file included from src/util/areastore.h:31,
                 from src/util/areastore.cpp:20:
5.4.0-447-g1dc1305ad/src/cmake_config.h:8: note: this is the location of the previous definition
 #define VERSION_MINOR 5
 
In file included from src/config.h:13,
                 from src/util/serialize.h:27,
                 from src/util/areastore.cpp:21:
src/cmake_config.h:9: warning: "VERSION_PATCH" redefined
 #define VERSION_PATCH 1
 
In file included from src/util/areastore.h:31,
                 from src/util/areastore.cpp:20:
5.4.0-447-g1dc1305ad/src/cmake_config.h:9: note: this is the location of the previous definition
 #define VERSION_PATCH 0
 
In file included from src/config.h:13,
                 from src/util/serialize.h:27,
                 from src/util/areastore.cpp:21:
src/cmake_config.h:11: warning: "VERSION_STRING" redefined
 #define VERSION_STRING "5.4.1"
 
In file included from src/util/areastore.h:31,
                 from src/util/areastore.cpp:20:
5.4.0-447-g1dc1305ad/src/cmake_config.h:11: note: this is the location of the previous definition
 #define VERSION_STRING "5.5.0-dev"
 
In file included from src/config.h:13,
                 from src/util/serialize.h:27,
                 from src/util/areastore.cpp:21:
src/cmake_config.h:12: warning: "PRODUCT_VERSION_STRING" redefined
 #define PRODUCT_VERSION_STRING "5.4"
 
In file included from src/util/areastore.h:31,
                 from src/util/areastore.cpp:20:
5.4.0-447-g1dc1305ad/src/cmake_config.h:12: note: this is the location of the previous definition
 #define PRODUCT_VERSION_STRING "5.5"
 
In file included from src/config.h:13,
                 from src/util/serialize.h:27,
                 from src/util/areastore.cpp:21:
src/cmake_config.h:13: warning: "STATIC_SHAREDIR" redefined
 #define STATIC_SHAREDIR "."
 
In file included from src/util/areastore.h:31,
                 from src/util/areastore.cpp:20:
5.4.0-447-g1dc1305ad/src/cmake_config.h:13: note: this is the location of the previous definition
 #define STATIC_SHAREDIR "/usr/local/share/minetest"
 
In file included from src/config.h:13,
                 from src/util/serialize.h:27,
                 from src/util/areastore.cpp:21:
src/cmake_config.h:14: warning: "STATIC_LOCALEDIR" redefined
 #define STATIC_LOCALEDIR "locale"
 
In file included from src/util/areastore.h:31,
                 from src/util/areastore.cpp:20:
5.4.0-447-g1dc1305ad/src/cmake_config.h:14: note: this is the location of the previous definition
 #define STATIC_LOCALEDIR "/usr/local/share/locale"
 
In file included from src/config.h:13,
                 from src/util/serialize.h:27,
                 from src/util/areastore.cpp:21:
src/cmake_config.h:16: warning: "ICON_DIR" redefined
 #define ICON_DIR "unix/icons"
 
In file included from src/util/areastore.h:31,
                 from src/util/areastore.cpp:20:
5.4.0-447-g1dc1305ad/src/cmake_config.h:16: note: this is the location of the previous definition
 #define ICON_DIR "/usr/local/share/icons"
 
In file included from src/config.h:13,
                 from src/util/serialize.h:27,
                 from src/util/areastore.cpp:21:
src/cmake_config.h:17: warning: "RUN_IN_PLACE" redefined
 #define RUN_IN_PLACE 1
 
In file included from src/util/areastore.h:31,
                 from src/util/areastore.cpp:20:
5.4.0-447-g1dc1305ad/src/cmake_config.h:17: note: this is the location of the previous definition
 #define RUN_IN_PLACE 0

```

**Comment by erlehmann** [comment]
@sfan5 @neoh4x0r I may just have very bad luck with build roulette. If you can not reproduce it *on x86*, I will not object to closing this issue – but otherwise please leave it open until some other x86 user can try it out.

**Comment by neoh4x0r** [comment]
> @sfan5 @neoh4x0r I may just have very bad luck with build roulette. If you can not reproduce it _on x86_, I will not object to closing this issue – but otherwise please leave it open until some other x86 user can try it out.

What version of Debian are you running?

I built minetest on 64-bit Debian 10 (Buster):
```
$ cat /etc/os-release | grep -i "PRETTY_NAME"
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
```

**Comment by erlehmann** [comment]
@neoh4x0r why is that relevant?

```
; cat /etc/debian_version 
10.10
```

**Comment by neoh4x0r** [comment]
> @neoh4x0r why is that relevant?
> 
> ```
> ; cat /etc/debian_version 
> 10.10
> ```

It would be relevant if you were trying to build on an old version of Debian.

However, that's no the case as we are both running buster - so it might just be an issue on i386.

When I have time I might be able to setup an i386 version of debian/buster and see if I can reproduce it.

**Comment by neoh4x0r** [comment]
**I can confirm that this bug applies to i386 only.**

I setup an i386 Debian Buster VM and compiled minetest.

I was able to reproduce the failure on the original commit 1dc1305ada07da8c2a278b46a34d58af86184af9
and the latest commit 80c3c7e642749f9316a3eee2c235df3ce8be1666

```
=============================
i386/Release build for 
 (master@ 1dc1305ada07da8c2a278b46a34d58af86184af9)
 (master@ 80c3c7e642749f9316a3eee2c235df3ce8be1666)
=============================

Test assertion failed: readF1000(is) == 53.534f
    at test_serialiation.cpp:233

=============================
i386/Debug build for 
 (master@ 1dc1305ada07da8c2a278b46a34d58af86184af9)
 (master@ 80c3c7e642749f9316a3eee2c235df3ce8be1666)
=============================

Test assertion failed: readF1000(is) == 53.534f
    at test_serialiation.cpp:233

Test assertion failed: !memcmp(&data[0], test_serialized_sta, data.size())
    at test_serialiation.cpp:280

```

**Comment by erlehmann** [comment]
> I can confirm that this bug applies to i386 only.

Nitpicking, but what you can confirmed that this bug applies to i386 – but not that it applies to i386 *only*.

Thank you very much for the reproduction!

**Comment by neoh4x0r** [comment]
> > I can confirm that this bug applies to i386 only.
> 
> Nitpicking, but what you can confirmed that this bug occurs on i386 – but not that it occurs on i386 _only_.
> 
> Thank you very much for the reproduction!

Actually I can confirm that.

I compiled a 64-bit version (on Debian 10 buster) and the problem did not occur.
This tells me that the problem only affects i386.

**Comment by erlehmann** [comment]
> Actually I can confirm that.

Minetest runs on more architectures than x86 and x86_64 though.

**Comment by erlehmann** [comment]
> It might also simply be a problem with floating-point precision 64-bit vs 32-bit.

I see no *legit* reason why floating point calculations should have a different precision on different platforms – but if they do, that is a bug and has to be fixed.

**Comment by neoh4x0r** [comment]
> > It might also simply be a problem with floating-point precision 64-bit vs 32-bit.
> 
> I see no _legit_ reason why floating point calculations should have a different precision on different platforms – but if they do, that is a bug and has to be fixed.

It's not about calculating floating-point values -- it's about comparing them (which requires an epsilon value).
It's the comparison that is failing in the test.
```
if(abs(a - b) < EPSILON) {
 // they are equal
} else {
 // not equal (eg. abs(a - b) >= EPSILON)
}
```

**Comment by erlehmann** [comment]
> It's not about calculating floating-point values -- it's about comparing them (which requires an epsilon value).

I have heard that a lot, but it depends on the use case and often it is not necessary at all to do it that way.

Regardless, why would such an epsilon variable have a different value depending on the platform?

**Comment by neoh4x0r** [comment]
> > It's not about calculating floating-point values -- it's about comparing them (which requires an epsilon value).
> 
> I have read that a lot, but it depends on the use case – often it is not necessary at all to do it that way.
> 
> Regardless, why would such an epsilon variable have a different value depending on the platform?

There is a lot of literature about the usage of a "machine epsilon" value (which is specific to particular floating-point unit) to determine if the difference between two floats are close enough (less than epsilon) that they are considered to be equal.

According to Wikipedia (https://en.wikipedia.org/wiki/Machine_epsilon#Formal_definition)

> Rounding is a procedure for choosing the representation of a real number in a floating point number system.
> For a number system and a rounding procedure, machine epsilon is the maximum relative error of the chosen rounding procedure.

This (at least to me) implies that there isn't one **"machine epsilon"** value that will work everywhere.
The short of it -- is that it is somewhat complicated (eg. rounding errors which can vary from one platform/device to the next).

**Comment by erlehmann** [comment]
I have bad news.

For me, the following program had different results on x86 and x86_64, compiled with GCC:

```
#include <cstdio>
int main() {
  double x = 50178230318.0;
  double y = 100000000000.0;
  printf( "%.16f", x / y );
  printf( "\n" );
  return 0;
}
```

On x86 it prints `0.5017823031799999`, on x86_64 it prints `0.5017823031800001`.

https://lemire.me/blog/2020/06/26/gcc-not-nearest/

**Comment by erlehmann** [comment]
I have tried out a tip from https://gcc.gnu.org/bugzilla/show_bug.cgi?id=35488#c4

> The i?86 FPU computes values in 80bit precision by default.  Starting with
GCC 4.3 you can use -mpc64 to reduce its precision to 64bits thereby obtaining
the valid IEEE-754 result.

With `g++ test.cpp '-mpc64' && ./a.out` I get `0.5017823031800001`.

Seems to me that one should add `-mpc64` to the GCC x86 build options.

**Comment by neoh4x0r** [comment]
> I have tried out a tip from https://gcc.gnu.org/bugzilla/show_bug.cgi?id=35488#c4
> 
> > The i?86 FPU computes values in 80bit precision by default.  Starting with
> > GCC 4.3 you can use -mpc64 to reduce its precision to 64bits thereby obtaining
> > the valid IEEE-754 result.
> 
> With `g++ test.cpp '-mpc64' && ./a.out` I get `0.5017823031800001`.
> 
> Seems to me that someone should add `-mpc64` to the build options (if this fixes the bug).

I exported **CXXFLAGS=-mpc64** before compiling and set VERBOSE=1 when running make.
And -mpc64 was added during the compilation.

```
$ export CXXFLAGS=-mpc64
$ cmake -S . -B build/linux -DRUN_IN_PLACE=1 \
-DCMAKE_BUILD_TYPE=Debug
$ VERBOSE=1 make -j$(nproc) -C build/linux -f Makefile
```

However, this had no bearing on the unittest failures.

I believe that test data is actually incorrect.

As the floating-point hex represention is not correct.

```
union float_bits {
	unsigned int i;
	float f;
} bits;


float float_values[4] = {
	53.534f,
	-300000.32f,
	F1000_MIN,
	F1000_MAX
};

for(int i=0;i<4;i++) {
	bits.f = float_values[i];
	fprintf(stderr,"%f (f) = 0x%08X\n", bits.f, bits.i);
}
```

The result of this was:
```
53.534000 (f) = 0x425622D1
-300000.312500 (f) = 0xC8927C0A
-2147483.000000 (f) = 0xCA03126C
2147483.000000 (f) = 0x4A03126C
```

The relavent part of the test data has this:

```
// index 30: 53.534[67]f
0x00, 0x00, 0xd1, 0x1e,

// index 34: -300000.32f
0xee, 0x1e, 0x5b, 0xc0,

// index 38: F1000_MIN
0x80, 0x00, 0x02, 0x80,

// index 42: F1000_MAX
0x7F, 0xFF, 0xFD, 0x80,
```

Moreover, on read-test compares to 53.534f, but 
the write-test uses 53.53467f

I think these descrepencies should be addressed before trying
to figure out another reason for test failures.

**Comment by erlehmann** [comment]
Possibly related: The bug where the engine is able to emit a float that it does not accept. https://github.com/minetest/minetest/issues/11742

**Comment by neoh4x0r** [comment]
> Possibly related: The bug where the engine is able to emit a float that it does not accept. #11742

I think the repeated casting from float to s32/u32 causes more problems than it solves.

We should probably be storing the raw ieee-745 float using **src/util/ieee_float.cpp** (u32Tof32Slow and f32Tou32Slow) to serialize the floats
and do away with the direct conversion to u32 and FIXED-POINT.

I see that those functions contain the word **slow**, but I'm sure that there are faster algorithms to do this.


In the following code snippets, the data that gets serialized for F1000 is the fixed-point converted value (cast to a s32 and u32).

For instance: F1000_MIN: ((float)(s32)((float)(-0x7FFFFFFF - 1) / FIXEDPOINT_FACTOR)) 
```
(
	(float)(s32)(
		(float) (-0x7FFFFFFF - 1) 
		/ FIXEDPOINT_FACTOR
	)
)

All of that casting is unnecessary.

// (float)0x80000000 = 2147483.648f / FIXEDPOINT_FACTOR = 2147483

// just define a constant float value
#define F1000_MIN -2147483f
#define F1000_MAX +2147483f
```


**src/unittest/test_serialization.cpp**
```
void TestSerialization::testStreamRead()
[..]
    UASSERT(readF1000(is) == -300000.32f);
[..]

void TestSerialization::testStreamWrite()
[..]
    writeF1000(os, -300000.32f);
[..]

```

**src/util/serialize.h**
```
#define FIXEDPOINT_FACTOR 1000.0f

// -0x7FFFFFFF - 1 -- just set this to 0x80000000
#define F1000_MIN ((float)(s32)((float)(-0x7FFFFFFF - 1) / FIXEDPOINT_FACTOR))
// 0x7FFFFFFF -- just set this to 0x80000000-1
#define F1000_MAX ((float)(s32)((float)(0x7FFFFFFF) / FIXEDPOINT_FACTOR))

inline f32 readF1000(const u8 *data)
{
	return (f32)readS32(data) / FIXEDPOINT_FACTOR;
}

inline s32 readS32(const u8 *data)
{
	return (s32)readU32(data);
}

inline u32 readU32(const u8 *data)
{
	u32 val;
	memcpy(&val, data, 4);
	return be32toh(val);
}

inline void writeF1000(u8 *data, f32 i)
{
	assert(i >= F1000_MIN && i <= F1000_MAX);
	writeS32(data, i * FIXEDPOINT_FACTOR);
}

inline void writeS32(u8 *data, s32 i)
{
	writeU32(data, (u32)i);
}

inline void writeU32(u8 *data, u32 i)
{
	u32 val = htobe32(i);
	memcpy(data, &val, 4);
}

```

**Comment by erlehmann** [comment]
> I think the repeated casting from float to s32/u32 causes more problems than it solves.

I agree. Do the values even roundtrip correctly?

**Comment by neoh4x0r** [comment]
> > I think the repeated casting from float to s32/u32 causes more problems than it solves.
> 
> I agree. Do the values even roundtrip correctly?

Looking at the MS documentation: https://docs.microsoft.com/en-us/dotnet/api/system.single?view=net-6.0

> A value is said to round-trip if an operation converts an original floating-point number to another form, an inverse
> operation transforms the converted form back to a floating-point number, and the final floating-point number is equal 
> to the original floating-point number.

**So this process would be:**
1. Start with a specific float value (A)
2. Cast that to a u32 (Unsigned 32-bit integer)
3. Cast the u32 back to float (B)
4. Is B == A ? if not roundtrip didn't work.

Seems like that would be a **unit test** that needs to be added.

**Comment by neoh4x0r** [comment]
**TL;DR** Using SSE registers for floating point operations solves this issue.
---
So, I just got done testing out the proposed solution of using SSE for all floating point operations.
I had no failures (with the defaults) on Debian (Buster 10) x86_64 -- but on i386 the failures occurred.

With SSE support enabled, the unit tests no longer fail.

The question is whether the use of sse2 floating point math should be enabled by default (for all platforms) or
if it should only be enabled for specific platforms (eg. i386/686).

I don't know what other platforms may suffer from this issue.

Moreover, what platforms are in-use that do not have sse2 support (such that, for them, sse2 support would need to be disabled)?

---

**Enable support for using sse registers for fp operations (gets put in CMakeLists.txt, both irrlicht and minetest)**
```
# use SSE for floating-point operations to avoid issues with improper fp-rounding and loss of precision
# when moving fp-data to incompatible or less-precise registers/storage locations
# see https://gcc.gnu.org/wiki/FloatingPointMath and https://gcc.gnu.org/wiki/x87note

add_compile_options(-mfpmath=sse -msse2)
```

+ **Using compiler-defaults for fp register operations (on Debian i386)**  
    + **Release mode**  
Test assertion failed: readF1000(is) == 53.534f
at test_serialization.cpp:233
[FAIL] testStreamRead - 0ms
[PASS] testStreamWrite - 0ms

    + **Debug mode**  
Test assertion failed: readF1000(is) == 53.534f
at test_serialization.cpp:233
[FAIL] testStreamRead - 0ms
Test assertion failed: !memcmp(&data[0], test_serialized_data, data.size())
at test_serialization.cpp:280
[FAIL] testStreamWrite - 0ms

+ **Using sse for fp register operations (on Debian i386)**  
    + **Release mode** / **Debug mode**  
[PASS] testStreamRead - 0ms
[PASS] testStreamWrite - 0ms

**Comment by neoh4x0r** [comment]
I just opened two PRs (this also affects irrlicht)

**Irrlicht:** https://github.com/minetest/irrlicht/pull/83
**minetest:** https://github.com/minetest/minetest/pull/11853

**Comment by neoh4x0r** [comment]
From the PR #11853

> I own a relatively recent laptop (Reform 2 from 2021) that does not have SSE (ARM processor).

@erlehmann Are you actually running Debian/i386.
You cannot run i386 on arm (it must be either be armel / armhf / arm64)
Unless you are running in a virtual environment.

**What is the output of this command for you?**
```
$ cat /proc/cpuinfo  | grep -i "model name" | sort -u
model name	: Intel(R) Core(TM) i7-2600 CPU @ 3.40GHz
```

**Comment by erlehmann** [comment]
@neoh4x0r I own several computers, among them a Thinkpad T60 and an MNT Reform 2.

The computer on which I encountered this bug has the following output for ´</proc/cpuinfo grep -i 'model name' |sort -u´:

´´´
model name	: Intel(R) Core(TM) Duo CPU      T2400  @ 1.83GHz
´´´

**Comment by neoh4x0r** [comment]
> @neoh4x0r I own several computers, among them a Thinkpad T60 and an MNT Reform 2.
> 
> The computer on which I encountered this bug has the following output for `</proc/cpuinfo grep -i 'model name' |sort -u`:
> 
> ```
> model name	: Intel(R) Core(TM) Duo CPU      T2400  @ 1.83GHz
> ```

Does the problem occur on the Reform 2 (arm processor) ?

The proposed solution should solve the issue on your Thinkpad.

**Comment by erlehmann** [comment]
> Does the problem occur on the Reform 2 (arm processor) ?

I doubt it, since it does not even have an x87 FPU.

**Comment by neoh4x0r** [comment]
> > Does the problem occur on the Reform 2 (arm processor) ?
> 
> I doubt it, since it does not even have an x87 FPU. I can test some time later.

The lack of an x87 FPU doesn't necessarily mean it is not affected (just less likely).
It could affect any system that behaves similarly to the intel design of the x87 FPU.

**Comment by neoh4x0r** [comment]
The following PRs related to this issue have been merged.
https://github.com/minetest/irrlicht/pull/83
https://github.com/minetest/minetest/pull/11853

Please check if this issue has been fixed.

**Comment by erlehmann** [comment]
> Please check if this issue has been fixed.

It seems so, on my machine (x86) the output is now:

```
======== Testing module TestSerialization
[PASS] testSerializeString - 0ms
[PASS] testDeSerializeString - 0ms
[PASS] testSerializeLongString - 0ms
[PASS] testDeSerializeLongString - 5ms
[PASS] testSerializeJsonString - 1ms
[PASS] testStreamRead - 0ms
[PASS] testStreamWrite - 0ms
[PASS] testFloatFormat - 0ms
```
